### PR TITLE
feat!: Add support for custom time periods in budget module

### PR DESCRIPTION
**Pull request by abhisek00**
Fix for the #729

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
@abhisek00
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_documentation
 The documentation needs to be regenerated. Please run `make generate_docs`. 
 ```bash 
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/modules/budget/README.md /tmp/tmp.Cj7jjclnj2/generate_docs/workspace/modules/budget/README.md
27a28
> | calendar\_period | Specifies the calendar period for the budget. Possible values are MONTH, QUARTER, YEAR, CUSTOM, CALENDAR\_PERIOD\_UNSPECIFIED | `string` | n/a | yes |
29a31,32
> | custom\_period\_end\_date | Specifies the end date for the calendar\_period CUSTOM | <pre>object(<br>    {<br>      year    = number<br>      month   = number<br>      day     = number<br>    }<br>  )</pre> | <pre>{<br>  "day": 1,<br>  "month": 1,<br>  "year": 1<br>}</pre> | no |
> | custom\_period\_start\_date | Specifies the start date for the calendar\_period CUSTOM | <pre>object(<br>    {<br>      year    = number<br>      month   = number<br>      day     = number<br>    }<br>  )</pre> | <pre>{<br>  "day": 1,<br>  "month": 1,<br>  "year": 1<br>}</pre> | no |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
 ``` 
- ⚠️check_whitespace
 Failed whitespace check. More details below. 
 ```bash 
Checking for trailing whitespace
./modules/budget/main.tf:55:   
./modules/budget/main.tf:66:      
./modules/budget/main.tf:70:    
./modules/budget/main.tf:71:    
Error: Trailing whitespace found in the lines above.
Checking for missing newline at end of file
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
modules/budget/main.tf
--- old/modules/budget/main.tf
+++ new/modules/budget/main.tf
@@ -50,25 +50,25 @@
 
     dynamic "custom_period" {
 
-    for_each = var.calendar_period == "CUSTOM" ? [1] : []
-    content {
-   
+      for_each = var.calendar_period == "CUSTOM" ? [1] : []
+      content {
+
         start_date {
-          year = var.custom_period_start_date.year
+          year  = var.custom_period_start_date.year
           month = var.custom_period_start_date.month
-          day = var.custom_period_start_date.day
+          day   = var.custom_period_start_date.day
         }
         end_date {
-          year = var.custom_period_end_date.year
+          year  = var.custom_period_end_date.year
           month = var.custom_period_end_date.month
-          day = var.custom_period_end_date.day
+          day   = var.custom_period_end_date.day
         }
-      
-    }
+
+      }
 
     }
-    
-    
+
+
   }
 
   amount {
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
modules/budget/variables.tf
--- old/modules/budget/variables.tf
+++ new/modules/budget/variables.tf
@@ -62,15 +62,15 @@
   description = "Specifies the start date for the calendar_period CUSTOM"
   type = object(
     {
-      year    = number
-      month   = number
-      day     = number
+      year  = number
+      month = number
+      day   = number
     }
   )
   default = {
-    day =   1
+    day   = 1
     month = 1
-    year =  1
+    year  = 1
   }
 }
 
@@ -78,15 +78,15 @@
   description = "Specifies the end date for the calendar_period CUSTOM"
   type = object(
     {
-      year    = number
-      month   = number
-      day     = number
+      year  = number
+      month = number
+      day   = number
     }
   )
   default = {
-    day =   1
+    day   = 1
     month = 1
-    year =  1
+    year  = 1
   }
 }
 
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
 ```

**Comment by comment-bot-dev** [comment]
@abhisek00
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_whitespace
 Failed whitespace check. More details below. 
 ```bash 
Checking for trailing whitespace
./modules/budget/main.tf:55:   
./modules/budget/main.tf:66:      
./modules/budget/main.tf:70:    
./modules/budget/main.tf:71:    
Error: Trailing whitespace found in the lines above.
Checking for missing newline at end of file
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
modules/budget/main.tf
--- old/modules/budget/main.tf
+++ new/modules/budget/main.tf
@@ -50,25 +50,25 @@
 
     dynamic "custom_period" {
 
-    for_each = var.calendar_period == "CUSTOM" ? [1] : []
-    content {
-   
+      for_each = var.calendar_period == "CUSTOM" ? [1] : []
+      content {
+
         start_date {
-          year = var.custom_period_start_date.year
+          year  = var.custom_period_start_date.year
           month = var.custom_period_start_date.month
-          day = var.custom_period_start_date.day
+          day   = var.custom_period_start_date.day
         }
         end_date {
-          year = var.custom_period_end_date.year
+          year  = var.custom_period_end_date.year
           month = var.custom_period_end_date.month
-          day = var.custom_period_end_date.day
+          day   = var.custom_period_end_date.day
         }
-      
-    }
+
+      }
 
     }
-    
-    
+
+
   }
 
   amount {
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
modules/budget/variables.tf
--- old/modules/budget/variables.tf
+++ new/modules/budget/variables.tf
@@ -62,15 +62,15 @@
   description = "Specifies the start date for the calendar_period CUSTOM"
   type = object(
     {
-      year    = number
-      month   = number
-      day     = number
+      year  = number
+      month = number
+      day   = number
     }
   )
   default = {
-    day =   1
+    day   = 1
     month = 1
-    year =  1
+    year  = 1
   }
 }
 
@@ -78,15 +78,15 @@
   description = "Specifies the end date for the calendar_period CUSTOM"
   type = object(
     {
-      year    = number
-      month   = number
-      day     = number
+      year  = number
+      month = number
+      day   = number
     }
   )
   default = {
-    day =   1
+    day   = 1
     month = 1
-    year =  1
+    year  = 1
   }
 }
 
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
 ```

**Comment by comment-bot-dev** [comment]
@abhisek00
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_documentation
 The documentation needs to be regenerated. Please run `make generate_docs`. 
 ```bash 
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/modules/budget/README.md /tmp/tmp.MNBG3DssUj/generate_docs/workspace/modules/budget/README.md
28c28
< | calendar\_period | Specifies the calendar period for the budget. Possible values are MONTH, QUARTER, YEAR, CUSTOM, CALENDAR\_PERIOD\_UNSPECIFIED | `string` | n/a | yes |
---
> | calendar\_period | Specifies the calendar period for the budget. Possible values are MONTH, QUARTER, YEAR, CUSTOM, CALENDAR\_PERIOD\_UNSPECIFIED | `string` | `null` | no |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
 ``` 
- ⚠️check_whitespace
 Failed whitespace check. More details below. 
 ```bash 
Checking for trailing whitespace
./modules/budget/main.tf:55:   
./modules/budget/main.tf:66:      
./modules/budget/main.tf:70:    
./modules/budget/main.tf:71:    
Error: Trailing whitespace found in the lines above.
Checking for missing newline at end of file
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
modules/budget/main.tf
--- old/modules/budget/main.tf
+++ new/modules/budget/main.tf
@@ -50,25 +50,25 @@
 
     dynamic "custom_period" {
 
-    for_each = var.calendar_period == "CUSTOM" ? [1] : []
-    content {
-   
+      for_each = var.calendar_period == "CUSTOM" ? [1] : []
+      content {
+
         start_date {
-          year = var.custom_period_start_date.year
+          year  = var.custom_period_start_date.year
           month = var.custom_period_start_date.month
-          day = var.custom_period_start_date.day
+          day   = var.custom_period_start_date.day
         }
         end_date {
-          year = var.custom_period_end_date.year
+          year  = var.custom_period_end_date.year
           month = var.custom_period_end_date.month
-          day = var.custom_period_end_date.day
+          day   = var.custom_period_end_date.day
         }
-      
-    }
+
+      }
 
     }
-    
-    
+
+
   }
 
   amount {
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
modules/budget/variables.tf
--- old/modules/budget/variables.tf
+++ new/modules/budget/variables.tf
@@ -63,15 +63,15 @@
   description = "Specifies the start date for the calendar_period CUSTOM"
   type = object(
     {
-      year    = number
-      month   = number
-      day     = number
+      year  = number
+      month = number
+      day   = number
     }
   )
   default = {
-    day =   1
+    day   = 1
     month = 1
-    year =  1
+    year  = 1
   }
 }
 
@@ -79,15 +79,15 @@
   description = "Specifies the end date for the calendar_period CUSTOM"
   type = object(
     {
-      year    = number
-      month   = number
-      day     = number
+      year  = number
+      month = number
+      day   = number
     }
   )
   default = {
-    day =   1
+    day   = 1
     month = 1
-    year =  1
+    year  = 1
   }
 }
 
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
 ```

**Comment by comment-bot-dev** [comment]
@abhisek00
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_documentation
 The documentation needs to be regenerated. Please run `make generate_docs`. 
 ```bash 
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/modules/budget/README.md /tmp/tmp.7UQP3Dt1cP/generate_docs/workspace/modules/budget/README.md
31,32c31,32
< | custom\_period\_end\_date | Specifies the end date for the calendar\_period CUSTOM | <pre>object(<br>    {<br>      year    = number<br>      month   = number<br>      day     = number<br>    }<br>  )</pre> | <pre>{<br>  "day": 1,<br>  "month": 1,<br>  "year": 1<br>}</pre> | no |
< | custom\_period\_start\_date | Specifies the start date for the calendar\_period CUSTOM | <pre>object(<br>    {<br>      year    = number<br>      month   = number<br>      day     = number<br>    }<br>  )</pre> | <pre>{<br>  "day": 1,<br>  "month": 1,<br>  "year": 1<br>}</pre> | no |
---
> | custom\_period\_end\_date | Specifies the end date for the calendar\_period CUSTOM | <pre>object(<br>    {<br>      year  = number<br>      month = number<br>      day   = number<br>    }<br>  )</pre> | <pre>{<br>  "day": 1,<br>  "month": 1,<br>  "year": 1<br>}</pre> | no |
> | custom\_period\_start\_date | Specifies the start date for the calendar\_period CUSTOM | <pre>object(<br>    {<br>      year  = number<br>      month = number<br>      day   = number<br>    }<br>  )</pre> | <pre>{<br>  "day": 1,<br>  "month": 1,<br>  "year": 1<br>}</pre> | no |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
 ```

**Comment by comment-bot-dev** [comment]
@abhisek00
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

File: modules/budget/main.tf [hunk 1]
```diff
@@ -46,6 +46,29 @@ resource "google_billing_budget" "budget" {
     credit_types_treatment = var.credit_types_treatment
     services               = local.services
     labels                 = var.labels
+    calendar_period        = var.calendar_period != "CUSTOM" ? var.calendar_period : null
+
+    dynamic "custom_period" {
+
```
Comment (bharathkkb) [inline modules/budget/main.tf:52]:
```suggestion
```

File: modules/budget/main.tf [hunk 2]
```diff
@@ -46,6 +46,29 @@ resource "google_billing_budget" "budget" {
...
+      for_each = var.calendar_period == "CUSTOM" ? [1] : []
+      content {
+
```
Comment (bharathkkb) [inline modules/budget/main.tf:55]:
```suggestion
```

File: modules/budget/main.tf [hunk 3]
```diff
@@ -46,6 +46,29 @@ resource "google_billing_budget" "budget" {
...
+          year  = var.custom_period_start_date.year
+          month = var.custom_period_start_date.month
+          day   = var.custom_period_start_date.day
+        }
+        end_date {
+          year  = var.custom_period_end_date.year
+          month = var.custom_period_end_date.month
+          day   = var.custom_period_end_date.day
+        }
+
```
Comment (bharathkkb) [inline modules/budget/main.tf:66]:
```suggestion
```

File: modules/budget/main.tf [hunk 4]
```diff
@@ -46,6 +46,29 @@ resource "google_billing_budget" "budget" {
...
+      }
+
```
Comment (bharathkkb) [inline modules/budget/main.tf:68]:
```suggestion
```

File: modules/budget/main.tf [hunk 5]
```diff
@@ -46,6 +46,29 @@ resource "google_billing_budget" "budget" {
...
+    }
+
+
```
Comment (bharathkkb) [inline modules/budget/main.tf:71]:
```suggestion
```

File: modules/budget/main.tf [hunk 6]
```diff
@@ -46,6 +46,29 @@ resource "google_billing_budget" "budget" {
     credit_types_treatment = var.credit_types_treatment
     services               = local.services
     labels                 = var.labels
+    calendar_period        = var.calendar_period != "CUSTOM" ? var.calendar_period : null
```
Comment (bharathkkb) [inline modules/budget/main.tf:49]: Lets move `var.calendar_period == "CUSTOM"` to a boolean local and we can reuse it here and below.

File: modules/budget/variables.tf [hunk 7]
```diff
@@ -53,6 +53,44 @@ variable "services" {
...
 
+variable "calendar_period" {
+  description = "Specifies the calendar period for the budget. Possible values are MONTH, QUARTER, YEAR, CUSTOM, CALENDAR_PERIOD_UNSPECIFIED"
+  type        = string
+  default     = null
+}
+
+variable "custom_period_start_date" {
+  description = "Specifies the start date for the calendar_period CUSTOM"
+  type = object(
```
Comment (bharathkkb) [inline modules/budget/variables.tf:64]: Might be simpler to express this as `DD-MM-YYYY`

File: modules/budget/variables.tf [hunk 8]
```diff
@@ -53,6 +53,44 @@ variable "services" {
   default     = null
 }
 
+variable "calendar_period" {
```
Comment (bharathkkb) [inline modules/budget/variables.tf:56]: Lets plumb some of these inputs to the main module
https://github.com/terraform-google-modules/terraform-google-project-factory/blob/9410ca40050668535068968a1a18a3ecfda915d7/main.tf#L90

**Review by bharathkkb — COMMENTED** [review]
Thanks for the PR @abhisek00

**Comment by comment-bot-dev** [comment]
@abhisek00
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

File: main.tf [hunk 9]
```diff
@@ -99,6 +99,10 @@ module "budget" {
   monitoring_notification_channels = var.budget_monitoring_notification_channels
   display_name                     = var.budget_display_name != null ? var.budget_display_name : null
   labels                           = var.budget_labels
+  calendar_period                  = var.calendar_period
```
Comment (bharathkkb) [inline main.tf:102]: nit: lets prefix these vars used in root module with `budget_`, so `budget_calendar_period` etc

File: modules/budget/main.tf [hunk 10]
```diff
@@ -18,6 +18,9 @@ locals {
   project_name     = length(var.projects) == 0 ? "All Projects" : var.projects[0]
   display_name     = var.display_name == null ? "Budget For ${local.project_name}" : var.display_name
   all_updates_rule = var.alert_pubsub_topic == null && length(var.monitoring_notification_channels) == 0 ? [] : ["1"]
+  custom_period    = var.calendar_period == "CUSTOM" ? [1] : []
+  start_date       = split("-", var.custom_period_start_date)
```
Comment (bharathkkb) [inline modules/budget/main.tf:22]: Lets split only if `custom_period`

File: variables.tf [hunk 11]
```diff
@@ -59,6 +59,24 @@ variable "billing_account" {
...
+variable "calendar_period" {
+  description = "Specifies the calendar period for the budget. Possible values are MONTH, QUARTER, YEAR, CUSTOM, CALENDAR_PERIOD_UNSPECIFIED"
+  type        = string
+  default     = null
+}
+
+variable "custom_period_start_date" {
+  description = "Specifies the start date (DD-MM-YYYY) for the calendar_period CUSTOM"
+  type        = string
+  default     = "1-1-2017"
```
Comment (bharathkkb) [inline variables.tf:71]: Same as above

File: modules/budget/variables.tf [hunk 12]
```diff
@@ -53,6 +53,24 @@ variable "services" {
...
+  description = "Specifies the calendar period for the budget. Possible values are MONTH, QUARTER, YEAR, CUSTOM, CALENDAR_PERIOD_UNSPECIFIED"
```
Comment (bharathkkb) [inline modules/budget/variables.tf:57]: Let's call out that `custom_period_start_date` and `custom_period_end_date` must be set if `CUSTOM`

**Comment by comment-bot-dev** [comment]
@abhisek00
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Review by bharathkkb — APPROVED** [review]
A last request - I just noticed this relies on [provider version 4.28+](https://github.com/hashicorp/terraform-provider-google/blob/main/CHANGELOG.md#4280-july-11-2022)

We should also update 
https://github.com/terraform-google-modules/terraform-google-project-factory/blob/9410ca40050668535068968a1a18a3ecfda915d7/versions.tf#L19-L27
and
https://github.com/terraform-google-modules/terraform-google-project-factory/blob/9410ca40050668535068968a1a18a3ecfda915d7/modules/budget/versions.tf#L20-L25
to use `~> 4.28`

**Comment by comment-bot-dev** [comment]
@abhisek00
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by comment-bot-dev** [comment]
@abhisek00
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Pull request merged**